demo-projects/blog/index.js

```diff
@@ -41,9 +45,31 @@ const admin = new AdminUI(keystone, {
   ],
 });
 
-module.exports = {
-  staticRoute,
-  staticPath,
-  keystone,
-  admin,
-};
+const server = new WebServer(keystone, {
+  adminUI,
+  port,
+  static: {
+    [staticRoute]: staticPath,
+    [WysiwygField.staticRoute]: WysiwygField.staticRoute,
+  },
+  next: next({
```

Comment:
I don't get this bit, would this mean we would have an opinion on what keystone users should use to build their front end? I was under the impression that this was something that keystone-5 was explicitly not trying to do?

Reply:
I'm also confused why we would create an API such as this when exposing the express app enables all this (and unlimited other things) with less learning on behalf of the developer.

Reply:
Yeah I was surprised when I did this too. But I tried it out and... actually kind of liked it.

I don't mind supporting a few popular "golden paths" of which integrating a next.js app could be one. _I think_. There's an argument to be made against this though.

It's sort of OK if we keep the supported integrations to a reasonable minimum, they introduce no new dependencies, and we're abstracting boilerplate (this next support passes all those tests for me)

> I'm also confused why we would create an API such as this when exposing the express app enables all this

Partly because if you're using next, supporting this and static assets removes the need to customise the server entirely.

Also I'm not as sold that our current solution doesn't require learning. I _myself_ get confused about things like

- Is the session middleware included in the stack before the app is available for customisation?
- How do I add express middleware _before_ the Admin UI middleware? (e.g static assets should almost never come after session initialisation, but that's the only option we have currently)

... etc. Right now I think we're actually in a "worst of both worlds" scenario, in that what _could be config isn't, but developers don't get full control_.

(at least in our demo configurations)

Along the same theme, I'd like to make it _more_ explicit how the middleware is all being set up if you're not using config. i.e if you're going to completely configure your own server, you should probably be responsible for binding the Admin UI middleware to it in the same way you're responsible for binding the Next middleware in the current demo project.

Sorry if I'm coming across as critical of what we've got; I'm just having a hard time working through the mix of static configuration vs. imperative setup that we're currently doing.

Reply:
I agree that we need to have "golden paths" but I’m not sure if having next as one one of them is right because in actual production apps, people will probably deploy their front end independently of keystone.

Reply:
> Along the same theme, I'd like to make it more explicit how the middleware is all being set up if you're not using config.

Yes to this.

I'm not especially familiar/up to date with this area of Keystone right now so this may make no sense -- could we split the configuration stuff into a separate package? Ie. out of the box, the dev just gets the building blocks (server, session middleware, etc.) and has to put them together themselves, very explicitly, or they can pull in the  "server configurator" package and have a much "nicer" but less powerful interface to get started.

If it worked, this would make it really obvious that the building blocks can be used directly. If you wanted to understand how they were usually used, you'd only need to understand a small amount of code (the configurator).